cmd/thanos/compact.go

```diff
@@ -147,9 +137,22 @@ func runCompact(
 
 		g.Add(func() error {
 			return runutil.Repeat(5*time.Minute, ctx.Done(), func() error {
+				if err := sy.SyncMetas(ctx); err != nil {
+					level.Error(logger).Log("msg", "sync failed", "err", err)
+				}
+				if err := sy.GarbageCollect(ctx); err != nil {
+					level.Error(logger).Log("msg", "garbage collection failed", "err", err)
+				}
 				for _, g := range sy.Groups() {
 					if _, err := g.Compact(ctx, comp); err != nil {
 						level.Error(logger).Log("msg", "compaction failed", "err", err)
+						// The HaltError type signals that we hit a critical bug and should block
+						// for investigation.
+						if compact.IsHaltError(err) {
```

Comment:
I totally see the reason here, but I feel like this need to be automated a little bit.. I can see cases when people will start it and don't have (or remember to have) alert for halted compactor.. And from orchestration point of view it will be seen as healthy..

Not sure what is the best way to handle it, maybe halting that as an option? (configured by flag) or maybe just return error and fail everything. What do you think?

Reply:
From a debugging point of view this is the best course of action I think. Remember that it should[tm] not happen to begin with. But if it did and we just exit with an error, we'll get rescheduled and debugging gets harder.

A less severe alternative would be marking ourselves as unhealthy. But in that case k8s will also just kill us eventually and reschedule.

I get that it's a bit of a dirty approach. But we do log an error and we are setting a respective metric.
Once we are fully confident that the issue is resolved we can maybe just exit?

Reply:
Sure I understand it, but let's hide it behind some debug flag then.